local character = require(script.Parent.Character)

return function()
	character = character:new()

	describe(":UpdateHealth()", function()
		it("should update the health", function()
			character:UpdateHealth(50)
			expect(character.Health).to.equal(50)
		end)
		it("should respect the maxHealth property", function()
			expect(function()
				character:UpdateHealth(125)
			end).to.throw()
		end)
	end)

	describe(":Jump()", function()
		it("should jump with the default value", function()
			expect(function()
				character:Jump()
			end).never.to.throw()
		end)
		it("should jump with the specified value", function()
			expect(function()
				character:Jump(105)
			end).never.to.throw()
		end)
	end)
end
