local character = require(script.Parent.Character)

return function()
	local newCharacter = character.new()

	describe(":UpdateHealth()", function()
		it("should update the health", function()
			newCharacter:UpdateHealth(50)
			expect(newCharacter.Health).to.equal(50)
		end)
		it("should respect the maxHealth property", function()
			expect(function()
				newCharacter:UpdateHealth(125)
			end).to.throw()
		end)
	end)

	describe(":Jump()", function()
		it("should jump with the default value", function()
			expect(function()
				newCharacter:Jump()
			end).never.to.throw()
		end)
		it("should jump with the specified value", function()
			expect(function()
				newCharacter:Jump(105)
			end).never.to.throw()
		end)
	end)
end
