local environment = require(script.Parent.Environment)

return function()
	local newEnvironment = environment.new()

	describe(":Get()", function()
		it("should get an element", function()
			newEnvironment:AddOne(0, "B")
			expect(newEnvironment:Get("B")).to.equal(0)
		end)
	end)
	describe(":AddOne()", function()
		it("should add an element", function()
			newEnvironment:AddOne(0, "A")
			expect(newEnvironment:Get("A")).to.equal(0)
		end)
	end)
	describe(":HasOne()", function()
		it("should check an element", function()
			newEnvironment:RemoveOne("B")
			expect(newEnvironment:HasOne("A")).to.equal(true)
			expect(newEnvironment:HasOne("B")).never.to.equal(true)
		end)
	end)
	describe(":RemoveOne()", function()
		it("should remove an element", function()
			newEnvironment:RemoveOne("A")
			expect(newEnvironment:Get("A")).never.to.be.ok()
		end)
	end)
	describe(":Add()", function()
		it("should add one or more elements", function()
			newEnvironment:Add({ A = 1, B = 2, C = 3 })
			expect(newEnvironment:Get("C")).to.equal(3)
		end)
	end)
	describe(":Has()", function()
		it("should check one or more elements", function()
			local hasAll = newEnvironment:Has({ "A", "B", "C" })
			expect(hasAll).to.equal(true)
		end)
	end)
	describe(":Update()", function()
		it("should update one or more elements", function()
			newEnvironment:Update({ A = 3, B = 2, C = 1 })
			expect(newEnvironment:Get("A")).to.equal(3)
			expect(newEnvironment:Get("C")).to.equal(1)
		end)
	end)
	describe(":Remove()", function()
		it("should remove one or more elements", function()
			newEnvironment:Remove({ "A", "B", "C" })
			expect(newEnvironment:Get("A")).never.to.be.ok()
			expect(newEnvironment:Get("C")).never.to.be.ok()
		end)
	end)

	local function test(n)
		if not n then
			error("something is not working")
		end
	end

	describe(":SetProperty()", function()
		it("should set a property in every element", function()
			newEnvironment:AddOne({ N = 0 }, "A")
			newEnvironment:SetProperty("N", 1)

			expect(newEnvironment:Get("A").N).to.equal(1)
		end)
	end)
	describe(":Run()", function()
		it("should execute a function in every element", function()
			newEnvironment:AddOne("A", { Test = test })
			expect(function()
				newEnvironment:Run("Test", 1)
			end).never.to.throw()
		end)
	end)
end
