local Scene = require(script.Parent.Scene)
local Particle = require(script.Parent.Particle)
local PhysicalObject = require(script.Parent.PhysicalObject)

return function()
	local newScene = Scene.new()
	local newParticle = Particle.new()
	local newPhysicalObject = PhysicalObject.new()
	newPhysicalObject:SetScene(newScene)

	describe(":SetSubject()", function()
		it("should set the particle subject", function()
			newParticle:SetSubject(newPhysicalObject)
			newParticle:SetSubject(newPhysicalObject)
			expect(newParticle.Subject.Id).to.equal(newPhysicalObject.Id)
		end)
	end)
	describe(":SetAngle()", function()
		it("should set the particle range", function()
			local val = Vector2.new(1, 1)
			newParticle:SetAngle(val)
			expect(newParticle.Angle).to.equal(val)
			expect(function()
				newParticle:SetAngle(-1)
			end).to.throw()
		end)
	end)
	describe(":SetMaxRate()", function()
		it("should set the particle max rate", function()
			newParticle:SetMaxRate(500)
			expect(newParticle.MaxRate).to.equal(500)
			expect(function()
				newParticle:SetMaxRate(-1)
			end).to.throw()
			expect(function()
				newParticle:SetMaxRate(2000)
			end).to.throw()
		end)
	end)
	describe(":Emit()", function()
		it("should emit particles", function()
			task.defer(function()
				expect(function()
					newParticle:Emit(50)
				end).never.to.throw()
			end)
		end)
	end)
end
