local Scene = require(script.Parent.Scene)
local Particle = require(script.Parent.Particle)
local PhysicalObject = require(script.Parent.PhysicalObject)

return function()
	Scene = Scene:new()
	Particle = Particle:new({ Parent = Scene })
	PhysicalObject = PhysicalObject:new({ Parent = Scene })

	describe(":SetSubject()", function()
		it("should set the particle subject", function()
			Particle:SetSubject(PhysicalObject)
			expect(Particle.Subject.Id).to.equal(PhysicalObject.Id)
		end)
	end)
	describe(":SetAngle()", function()
		it("should set the particle range", function()
			local val = Vector2.new(1, 1)
			Particle:SetAngle(val)
			expect(Particle.Angle).to.equal(val)
			expect(function()
				Particle:SetAngle(-1)
			end).to.throw()
		end)
	end)
	describe(":SetMaxRate()", function()
		it("should set the particle max rate", function()
			Particle:SetMaxRate(500)
			expect(Particle.MaxRate).to.equal(500)
			expect(function()
				Particle:SetMaxRate(-1)
			end).to.throw()
			expect(function()
				Particle:SetMaxRate(2000)
			end).to.throw()
		end)
	end)
	describe(":Emit()", function()
		it("should emit particles", function()
			expect(function()
				Particle:Emit(50)
			end).never.to.throw()
		end)
	end)
end
