local sceneManager = require(script.Parent.SceneManager)
local scene = require(script.Parent.Parent.Classes.Public.Scene)

return function()
	scene = scene:new()
	describe(":Load()", function()
		it("should load a scene", function()
			sceneManager:Load(scene)
			expect(sceneManager.ActiveScenes[scene.Id].Id).to.equal(scene.Id)
		end)
	end)
	describe(":Unload()", function()
		it("should unload a scene", function()
			sceneManager:Unload(scene)
			expect(sceneManager.ActiveScenes[scene.Id]).never.to.be.ok()
		end)
	end)
	describe(":FindByName()", function()
		it("should load a scene", function()
			scene:SetName("MyScene")
			expect(sceneManager:FindByName("MyScene").Id).to.equal(scene.Id)
		end)
	end)
end
