local sceneManager = require(script.Parent.SceneManager)
local scene = require(script.Parent.Parent.Classes.Public.Scene)

return function()
	local newScene = scene.new({ Parent = workspace })
	describe(":Load()", function()
		it("should load a scene", function()
			sceneManager:Load(newScene)
			expect(sceneManager.ActiveScenes[newScene.Id].Id).to.equal(newScene.Id)
		end)
	end)
	describe(":Unload()", function()
		it("should unload a scene", function()
			sceneManager:Unload(newScene)
			expect(sceneManager.ActiveScenes[newScene.Id]).never.to.be.ok()
		end)
	end)
	describe(":FindByName()", function()
		it("should load a scene", function()
			newScene:SetName("MyScene")
			expect(sceneManager:FindByName("MyScene").Id).to.equal(newScene.Id)
		end)
	end)
end
