local data = require(script.Parent.Parent.Parent.AppData.Data)
local baseClass = require(script.Parent.BaseClass)

return function()
	describe("constructor .new()", function()
		it("should create a new object", function()
			local object = baseClass.new()
			expect(data.objects[object.Id]).to.be.ok()
		end)
	end)

	describe(":SetName()", function()
		it("should rename the object", function()
			local object = baseClass.new()
			object:SetName("NewName")
			expect(object.Name).to.equal("NewName")
		end)
	end)

	describe(":Destroy()", function()
		it("Destroys the object", function()
			expect(function()
				local object = baseClass.new()
				object:Destroy()
			end)
		end)
	end)
end
